src/librustc/traits/select.rs

```diff
@@ -780,10 +777,11 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
         debug!("evaluate_trait_predicate_recursively({:?})",
                obligation);
 
-        if !self.intercrate.is_some() && obligation.is_global() {
-            // If a param env is consistent, global obligations do not depend on its particular
-            // value in order to work, so we can clear out the param env and get better
-            // caching. (If the current param env is inconsistent, we don't care what happens).
+        if !self.intercrate.is_some() && obligation.is_global()
+            && !obligation.param_env.is_global() {
+            // If a param env has no global bounds, global obligations do not
+            // depend on its particular value in order to work, so we can clear
+            // out the param env and get better caching.
```

Comment:
Well, I guess this is part of how we avoid the soundness problems too. I am trying to decide if this `is_global` check on the param environment suffices. I think it probably does given the current definitions of elaboration, but won't once we have a better definition of implied bounds (but that will require the newer solver, in which this PR is sort of easier to implement anyway).

I am imagining cases like this:

```rust
trait Foo where u32: Bar { }
trait Bar { }
```

Now, if we were to *fully* elaborate, `T: Foo` would imply `u32: Bar`.

Reply:
> Well, I guess this is part of how we avoid the soundness problems too.

Yes, but this only seemed to cause bounds to not apply in my testing.

> I am imagining cases like this:
> 
> ```rust
> trait Foo where u32: Bar { }
> trait Bar { }
> ```
>
> Now, if we were to fully elaborate, T: Foo would imply u32: Bar.

True, but this is surely just as much a problem with fully elaborating

> ```rust
> trait Foo where Vec<T>: Copy { }
> trait Bar { }
> ```